File: tests/test_doom_weapons.py
import unittest

from doom_weapons import WEAPON_TYPES, PlayerWeapon


class PistolExampleTest(unittest.TestCase):
    """weapons.md의 '## 예제: 권총 발사 한 번'을 그대로 옮긴 테스트."""

    def test_fire_cycles_through_attack_and_flash_back_to_ready(self):
        pistol = WEAPON_TYPES["pistol"]
        w = PlayerWeapon(pistol)
        self.assertIs(w.state, pistol.readyState)

        w.setState(pistol.attackState)
        for _ in range(pistol.attackState.duration):
            w.tick()
        self.assertIs(w.state, pistol.flashState)

        for _ in range(pistol.flashState.duration):
            w.tick()
        self.assertIs(w.state, pistol.readyState)


class WeaponRosterTest(unittest.TestCase):
    """weapons.md의 '## 전체 무기 로스터' 표 검증 — 탄약 종류·소비량·피해 공식."""

    def test_all_eight_weapons_present(self):
        expected = {"fist", "chainsaw", "pistol", "shotgun", "chaingun",
                    "rocketLauncher", "plasmaRifle", "bfg9000"}
        self.assertEqual(set(WEAPON_TYPES.keys()), expected)

    def test_fist_and_chainsaw_need_no_ammo(self):
        self.assertIsNone(WEAPON_TYPES["fist"].ammoType)
        self.assertIsNone(WEAPON_TYPES["chainsaw"].ammoType)

    def test_bullet_weapons_use_5_10_15_formula(self):
        for name in ("pistol", "chaingun"):
            wt = WEAPON_TYPES[name]
            for r in range(3):
                self.assertIn(wt.damage(lambda v=r: v), (5, 10, 15))

    def test_shotgun_fires_seven_pellets(self):
        # 펠릿 7발 * 최소 5 ~ 최대 15 사이.
        dmg = WEAPON_TYPES["shotgun"].damage(lambda: 0)
        self.assertEqual(dmg, 7 * 5)
        dmg_max = WEAPON_TYPES["shotgun"].damage(lambda: 2)
        self.assertEqual(dmg_max, 7 * 15)

    def test_bfg_costs_40_cells(self):
        self.assertEqual(WEAPON_TYPES["bfg9000"].ammoType, "cells")
        self.assertEqual(WEAPON_TYPES["bfg9000"].ammoPerShot, 40)

    def test_rocket_launcher_uses_rockets(self):
        self.assertEqual(WEAPON_TYPES["rocketLauncher"].ammoType, "rockets")
        self.assertEqual(WEAPON_TYPES["rocketLauncher"].ammoPerShot, 1)

    def test_bfg_hits_hardest_at_max_roll(self):
        bfg = WEAPON_TYPES["bfg9000"].damage(lambda: 700)
        rocket = WEAPON_TYPES["rocketLauncher"].damage(lambda: 132)
        self.assertGreater(bfg, rocket)


if __name__ == "__main__":
    unittest.main()

File: tests/cpp/test_doom_weapons.cpp
// examples/doom/weapons.md 검증 (C++17).
#include "doom_weapons.hpp"
#include "minitest.hpp"

using namespace Doom;

TEST(pistol_fire_cycles_through_attack_and_flash_back_to_ready) {
    auto pistol = weaponTypes().at("pistol");
    PlayerWeapon w(pistol);
    CHECK(w.state == pistol->readyState);

    w.setState(pistol->attackState);
    for (int i = 0; i < pistol->attackState->duration; i++) w.tick();
    CHECK(w.state == pistol->flashState);

    for (int i = 0; i < pistol->flashState->duration; i++) w.tick();
    CHECK(w.state == pistol->readyState);
}

TEST(weapon_roster_has_all_8) {
    static const char* expected[] = {"fist", "chainsaw", "pistol", "shotgun",
                                      "chaingun", "rocketLauncher", "plasmaRifle", "bfg9000"};
    CHECK_EQ(weaponTypes().size(), 8u);
    for (auto name : expected) CHECK(weaponTypes().count(name) == 1);
}

TEST(fist_and_chainsaw_need_no_ammo) {
    CHECK(weaponTypes().at("fist")->ammoType.empty());
    CHECK(weaponTypes().at("chainsaw")->ammoType.empty());
}

TEST(bullet_weapons_use_5_10_15_formula) {
    for (auto name : {"pistol", "chaingun"}) {
        auto wt = weaponTypes().at(name);
        for (int r = 0; r < 3; r++) {
            int dmg = wt->damage([r] { return r; });
            CHECK(dmg == 5 || dmg == 10 || dmg == 15);
        }
    }
}

TEST(shotgun_fires_seven_pellets) {
    auto shotgun = weaponTypes().at("shotgun");
    CHECK_EQ(shotgun->damage([] { return 0; }), 7 * 5);
    CHECK_EQ(shotgun->damage([] { return 2; }), 7 * 15);
}

TEST(bfg_costs_40_cells) {
    auto bfg = weaponTypes().at("bfg9000");
    CHECK_EQ(bfg->ammoType, "cells");
    CHECK_EQ(bfg->ammoPerShot, 40);
}

TEST(bfg_hits_hardest_at_max_roll) {
    int bfg = weaponTypes().at("bfg9000")->damage([] { return 700; });
    int rocket = weaponTypes().at("rocketLauncher")->damage([] { return 132; });
    CHECK(bfg > rocket);
}

TEST_MAIN()
